clippy_utils/src/higher.rs

```diff
@@ -428,12 +428,31 @@ pub fn extract_assert_macro_args<'tcx>(e: &'tcx Expr<'tcx>) -> Option<Vec<&'tcx
     /// compared
     fn ast_matchblock(matchblock_expr: &'tcx Expr<'tcx>) -> Option<Vec<&Expr<'_>>> {
         if_chain! {
-            if let ExprKind::Match(headerexpr, _, _) = &matchblock_expr.kind;
+            if let ExprKind::Match(headerexpr, arms, _) = &matchblock_expr.kind;
             if let ExprKind::Tup([lhs, rhs]) = &headerexpr.kind;
             if let ExprKind::AddrOf(BorrowKind::Ref, _, lhs) = lhs.kind;
             if let ExprKind::AddrOf(BorrowKind::Ref, _, rhs) = rhs.kind;
             then {
-                return Some(vec![lhs, rhs]);
+                let mut vec_arg = vec![lhs, rhs];
+                if_chain! {
+                    if !arms.is_empty();
+                    if let ExprKind::Block(Block{expr: Some(if_expr),..},_) = arms[0].body.kind;
+                    if let ExprKind::If(_, if_block, _) = if_expr.kind;
+                    if let ExprKind::Block(Block{stmts: stmts_if_block,..},_) = if_block.kind;
+                    if stmts_if_block.len() >= 2;
+                    if let StmtKind::Expr(call_assert_failed)
+                        | StmtKind::Semi(call_assert_failed) = stmts_if_block[1].kind;
+                    if let ExprKind::Call(_, args_assert_failed) = call_assert_failed.kind;
+                    if args_assert_failed.len() >= 4;
+                    if let ExprKind::Call(_, args) =  args_assert_failed[3].kind;
+                    if !args.is_empty();
+                    if let Some (mut format_arg_expn) = FormatArgsExpn::parse(&args[0]);
+                    then {
+                        vec_arg.push(format_arg_expn.format_string);
+                        vec_arg.append(&mut format_arg_expn.value_args);
```

Comment:
I think we should change `extract_assert_macro_args` to be more consistent with other utils here. We could have `AssertExpn::parse` and `AssertExpn` is a struct that contains a `FormatArgsExpn`.

Reply:
Thanks I have done something similar, but now I would like to test parsing assert(expr, "format {}", value), how can I test that ?

Reply:
Use Expand macros in the rust playground to see what you are trying to parse. For other tips, use `dbg!()` to set a checkpoint in your code and see if that line is being reached. Use `dbg!(expr)` to see the HIR. Use `let _ = dbg!();` in `if_chain!`s.